### PR TITLE
fix: load policy in compute dtype when optimizer holds fp32 master weights

### LINKED ISSUE #2865: [Automodel]: Hard-coded fp32 model weights doubles memory with FusedAdam
## Summary

The automodel backend hard-codes the model load dtype to **fp32** in order to support master weights [here](https://github.com/NVIDIA-NeMo/RL/blob/1ff6e117a41e37fdd80401ae375ec9f23e8d28ea/nemo_rl/models/automodel/setup.py#L305). 
However, the optimizer is configurable and supports both `torch.optim.AdamW` and TE `FusedAdam`. When **FusedAdam** is used, the fp32 load is unnecessary and causes **~2× optimizer master-weight memory** (and ~2× model-weight memory), because FusedAdam already maintains its own fp32 master internally.

## Details

With the TE FusedAdam optimizer, the fp32 master is kept **inside the optimizer**, via:

```yaml
optimizer:
  name: transformer_engine.pytorch.optimizers.fused_adam.FusedAdam
  kwargs:
    master_weights: true
    store_param_remainders: true
```

`store_param_remainders: true` lets FusedAdam split the fp32 master into **bf16 (high 16 bits) + int16 remainder (low 16 bits)** — `16 + 16 = 32 = fp32`. This depends on the parameters being **bf16**: once a parameter is fp32, TE detects `param.dtype != bf16` and **skips the remainder path, storing a separate full fp32 master** instead. Ref: [https://github.com/NVIDIA/TransformerEngine/blob/42b840051647eef89761a16dfdff87e82bb253ab/transformer_engine/pytorch/optimizers/fused_adam.py#L97](https://github.com/NVIDIA/TransformerEngine/blob/42b840051647eef89761a16dfdff87e82bb253ab/transformer_engine/pytorch/optimizers/fused_adam.py#L97)

So when the load dtype is hard-coded to fp32, with FusedAdam:

- **Model weights**: bf16 → fp32 = **2× memory**.
- **Optimizer master**: `bf16 + int16 remainder` → **full fp32** = **2× memory**.

## Reproduction

Run a grpo training, toggling only the model load dtype (`AutoConfig(torch_dtype=...)` in `nemo_rl/models/automodel/setup.py`) between
`float32` and `bfloat16`, and sum the `optimizer.state` tensor bytes after the first `optimizer.step()`.

## Observed (rank 0)

| metric                    | fp32 load (current)              | modified to bf16 load |
| ------------------------- | -------------------------------- | -------------------- |
| `master_param`            | **fp32 7.32 GB** + int16 0.60 GB | int16 — 4.26 GB      |
| optimizer state /rank     | **16.43 GB**                     | 12.77 GB             |
| optimizer state all-ranks | 131.5 GB                         | 102.2 GB             |
| model params all-ranks    | 66.6 GB                          | 35.7 GB              |

`exp_avg` / `exp_avg_sq` are unchanged (bf16); the entire increase is the master copy.

## Suggested fix

Choose the model load dtype based on the optimizer (AdamW vs FusedAdam) instead of hard-coding fp32: load fp32 only for optimizers without an internal master, and keep the compute dtype (bf16) when FusedAdam holds the master.

For reference, I have a candidate fix in: [https://github.com/NVIDIA-NeMo/RL/commit/e968d90f72925637e8834f81510b5f7a181b4093](https://github.com/NVIDIA-NeMo/RL/commit/e968d90f72925637e8834f81510b5f7a181b4093)

### PR DESCRIPTION
**Pull request by Doondi-Ashlesh**
# What does this PR do ?

Avoid doubling model- and optimizer-state memory on the AutoModel backend by loading the policy in fp32 only when the optimizer has no internal master, instead of always loading fp32.

# Issues

Closes #2865

# Details

The AutoModel backend always loaded the policy with `torch_dtype=float32` to preserve optimizer master weights. When the optimizer already keeps its own fp32 master (TE FusedAdam with `master_weights`, stored compactly as bf16 params plus an int16 remainder), the fp32 load duplicates that master and roughly doubles both model- and optimizer-state memory.

This PR selects the load dtype based on the optimizer:

* Optimizers that hold their own master (FusedAdam with `master_weights`): load in the compute dtype (e.g. bf16) and let the optimizer own the master.
* Masterless optimizers (e.g. `torch.optim.AdamW`): keep the fp32 load so updates do not underflow.

The Automodel checkpoint-dtype-restore workaround is now applied only on the fp32 path, where it is actually needed.

Reported memory impact from the issue (FusedAdam recipe):

| Metric | fp32 load | compute-dtype load |
| --- | --- | --- |
| Master params | 7.32 GB | 4.26 GB |
| Optimizer state per rank | 16.43 GB | 12.77 GB |
| Optimizer state all ranks | 131.5 GB | 102.2 GB |
| Model params all ranks | 66.6 GB | 35.7 GB |

# Usage

No config or API change. Existing recipes that use FusedAdam with `master_weights: true` automatically load in the compute dtype; AdamW recipes are unchanged.

# Additional Information
* New unit tests in `tests/unit/models/automodel/test_automodel_setup.py` cover the optimizer detection helper, the dtype selection in `validate_and_prepare_config`, and the restore gating in `setup_model_and_optimizer`.
* The unit tests require `torch` and `nemo_automodel`, so please trigger CI to run the full suite. The AdamW path is unchanged, so existing behavior and tests are preserved.

### CODE REVIEW

**Comment by Doondi-Ashlesh** [comment]
Hi, checking in on this PR, can a maintainer please review it. Happy to make further changes if required. Thanks!
